### PR TITLE
"SMS OTP authenticator modification added"

### PR DESCRIPTION
**Pull request by elilmatha**

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 1]
```diff
@@ -41,15 +41,13 @@
...
-import java.io.InputStream;
-import java.lang.System;
 import java.net.MalformedURLException;
 import java.net.ProtocolException;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
-import java.util.Properties;
+import java.io.*;
```
Comment (RKathees) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Import only specific classes.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 2]
```diff
@@ -192,14 +157,68 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
         if (userToken.equals(contextToken)) {
-            context.setSubject(AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier("an authorised user"));
+            context.setSubject(AuthenticatedUser
+                    .createLocalAuthenticatedUserFromSubjectIdentifier("an authorised user"));
+        } else if (savedOTPString != null && savedOTPString.contains(userToken)) {
+            context.setSubject(AuthenticatedUser
+                    .createLocalAuthenticatedUserFromSubjectIdentifier("an authorised user"));
+
+            savedOTPString = savedOTPString.replaceAll(userToken, "").replaceAll(",,", ",");
+
```
Comment (RKathees) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Remove empty line.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 3]
```diff
@@ -192,14 +157,68 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
 
     /**
+     * Get the user realm of the logged in user
+     */
+    private UserRealm getUserRealm(String username) throws AuthenticationFailedException {
+        UserRealm userRealm = null;
+        try {
+            String tenantDomain = MultitenantUtils.getTenantDomain(username);
+            int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
```
Comment (RKathees) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Remove empty line.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 4]
```diff
@@ -228,71 +245,120 @@ public String getContextIdentifier(HttpServletRequest httpServletRequest) {
...
+            if (httpMethod.trim().toLowerCase().equals(SMSOTPConstants.GET_METHOD.toLowerCase())) {
+                connection.setRequestMethod(SMSOTPConstants.GET_METHOD);
+            } else if (httpMethod.trim().toLowerCase().equals(SMSOTPConstants.POST_METHOD.toLowerCase())) {
+                connection.setRequestMethod(SMSOTPConstants.POST_METHOD);
+                if (!payload.trim().isEmpty()) {
+                    payload = payload.replaceAll("\\$ctx.num", mobile).replaceAll("\\$ctx.msg", smsMessage + otpToken);
+                }
+
+                OutputStreamWriter writer = null;
```
Comment (RKathees) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Import the class java.io.OutputStreamWriter. This authenticator's build is failing.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 5]
```diff
@@ -80,104 +75,74 @@ public boolean canHandle(HttpServletRequest request) {
...
-            } catch (Exception e) {
-                throw new AuthenticationFailedException("Cannot find the user realm", e);
-            }
+            UserRealm userRealm = getUserRealm(username);
             username = MultitenantUtils.getTenantAwareUsername(String.valueOf(username));
             if (userRealm != null) {
                 try {
-                    mobile = userRealm.getUserStoreManager().getUserClaimValue(username, SMSOTPConstants.MOBILE_CLAIM, null).toString();
+                    mobile = userRealm.getUserStoreManager()
+                            .getUserClaimValue(username, SMSOTPConstants.MOBILE_CLAIM, null).toString();
```
Comment (RKathees) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Here, toString() is redundant

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 6]
```diff
@@ -80,104 +75,74 @@ public boolean canHandle(HttpServletRequest request) {
...
+                } catch (UserStoreException e) {
+                    throw new AuthenticationFailedException("Cannot find the user claim for mobile " + e.getMessage(),
+                            e);
+                }
+                try {
+                    savedOTPString = userRealm.getUserStoreManager()
+                            .getUserClaimValue(username, SMSOTPConstants.SAVED_OTP_LIST, null).toString();
```
Comment (RKathees) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Here, toString() is redundant

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 7]
```diff
@@ -228,71 +245,120 @@ public String getContextIdentifier(HttpServletRequest httpServletRequest) {
...
+                }
+
+                OutputStreamWriter writer = null;
+                try {
+                    writer = new OutputStreamWriter(connection.getOutputStream(), "UTF-8");
+                    writer.write(payload);
+                } catch (IOException e) {
+                    throw new AuthenticationFailedException("Error while posting payload message", e);
+                } finally {
+                    writer.close();
```
Comment (RKathees) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: It may cause nullPointerException

**Comment by elilmatha** [comment]
closing pr since changes are done with https://github.com/wso2-extensions/identity-outbound-auth-sms-otp/pull/3

**Pull request closed without merging**